Add post about stabilizing intra-doc links (#689)
rust-lang/blog.rust-lang.org: posts/inside-rust/2020-09-11-stabilizing-intra-doc-links.md

layout: post
title: "Intra-doc links close to stabilization"
author: Manish Goregaokar and Joshua Nelson
team: the rustdoc team <https://www.rust-lang.org/governance/teams/dev-tools#rustdoc>
---

# Intra-doc links close to stabilization

> We're excited to share that

Reviewer: Isn't punctuation missing here?

Reviewer 2: oh it seems like some of the text got dropped @jyn514

Author: Hmm, that's what I see on the HackMD:

![image](https://user-images.githubusercontent.com/23638587/93004545-ed258980-f515-11ea-96af-c42709e8855d.png)

Did you mean to write more?

Reviewer 2: Alright, I must have messed up. This should be something like "We're excited to share that intra-doc links are stabilizing soon!"

It turns out there was a simple problem with the implementation - it assumed
all items were in the current crate! Clearly, that's not always the case. [The fix][resolve-cross-crate] turned out to be easy enough that I could implement it as my first contribution to rustdoc.

_Note from Manish:_ Actually, the distinction between [`DefId`] and [`LocalDefId`] _didn't exist_ when we wrote the feature, and the code would only resolve paths based on the resolver's current internal scope (which can only be within the current crate, since that is the only scope information the resolver had at the time). However, over time the compiler [gained the ability][refactor-resolve] to store and query resolution scopes of dependencies. We never noticed, and continued to believe that there was a large piece of work blocking stabilization.

However, my solution had one small problem: on certain [carefully crafted inputs][macro-in-closure], it would crash:

```rust
#![feature(decl_macro)]
fn main() {
    || {
        macro m() {}
    };
}

> thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', /home/joshua/src/rust/src/librustc_hir/definitions.rs:358:9

Reviewer: Maybe put this error into another codeblock so the highlighting isn't weird?

layout: post
title: "Intra-doc links close to stabilization"
author: Manish Goregaokar and Joshua Nelson
team: the rustdoc team <https://www.rust-lang.org/governance/teams/dev-tools#rustdoc>
---

> 
> # Intra-doc links close to stabilization

Reviewer 3: I think this is handled by the front matter, no?

```suggestion
```

Reviewer 3: ![image](https://user-images.githubusercontent.com/37223377/93159161-711c7480-f6c2-11ea-8374-902aa8572a3b.png)

We're excited to share that intra-doc links are stabilizing soon!

> [Intra-doc links] are a feature of `rustdoc` that allow you to link to '[items]' - functions, types, and more - by their name, instead of a hard-coded url. This lets you have accurate links even if your types are [re-exported in a different module or crate][broken-string-links]. Here is a simple example:

Reviewer 3: ```suggestion
[Intra-doc links] are a feature of `rustdoc` that allow you to link to '[items]' - functions, types, and more - by their name, instead of a hard-coded URL. This lets you have accurate links even if your types are [re-exported in a different module or crate][broken-string-links]. Here is a simple example:
```

```rust

> /// Link to [`f()`]
> pub struct S;
> pub fn f() {}

Reviewer 3: I would suggest adding a blank line for more visual clarity.

```suggestion
/// Link to [`f()`]
pub struct S;

pub fn f() {}
```

```

> Intra-doc links have been around for a while, all the way back [since 2017][tracking-issue]! What changed recently is that they will be available on `stable` starting with Rust 1.48.0. We recommend that you switch your libraries to use intra-doc links, which will fix broken links for re-exported types and links to different crates. We hope to add support for automating this process with [`cargo fix`] in the future.

Reviewer 2: ```suggestion
Intra-doc links have been around for a while, all the way back [since 2017][tracking-issue]! They have been available on `nightly` without flags (and thus, on docs.rs), so you may be surprised to hear that they weren't yet stable. What's changing now is that they will be available on stable Rust, which also means we are more confident in the implementation and would strongly encourage their use. We recommend that you switch your libraries to use intra-doc links, which will fix broken links for re-exported types and links to different crates. We hope to add support for automating this process with [`cargo fix`] in the future.
```

> Intra-doc links have been around for a while, all the way back [since 2017][tracking-issue]! They have been available on `nightly` without flags (and thus, on docs.rs), so you may be surprised to hear that they weren't yet stable. What's changing now is that they will be available on stable Rust, which also means we are more confident in the implementation and would strongly encourage their use. We recommend that you switch your libraries to use intra-doc links, which will fix broken links for re-exported types and links to different crates. We hope to add support for automating this process with [`cargo fix`] in the future.

Reviewer 3: Link to docs.rs?

```suggestion
Intra-doc links have been around for a while, all the way back [since 2017][tracking-issue]! They have been available on `nightly` without flags (and thus, on [docs.rs](https://docs.rs)), so you may be surprised to hear that they weren't yet stable. What's changing now is that they will be available on stable Rust, which also means we are more confident in the implementation and would strongly encourage their use. We recommend that you switch your libraries to use intra-doc links, which will fix broken links for re-exported types and links to different crates. We hope to add support for automating this process with [`cargo fix`] in the future.
```

all items were in the current crate! Clearly, that's not always the case. [The fix][resolve-cross-crate] turned out to be easy enough that I could implement it as my first contribution to rustdoc.

_Note from Manish:_ Actually, the distinction between [`DefId`] and [`LocalDefId`] _didn't exist_ when we wrote the feature, and the code would only resolve paths based on the resolver's current internal scope (which can only be within the current crate, since that is the only scope information the resolver had at the time). However, over time the compiler [gained the ability][refactor-resolve] to store and query resolution scopes of dependencies. We never noticed, and continued to believe that there was a large piece of work blocking stabilization.

However, my solution had one small problem: on certain [carefully crafted inputs][macro-in-closure], it would crash:

```rust
#![feature(decl_macro)]
fn main() {
    || {
        macro m() {}
    };
}
```
```
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', /home/joshua/src/rust/src/librustc_hir/definitions.rs:358:9
```

## HirIds and DefIds and trees, oh my!

> (If you're not interested in the internals of the rust compiler, feel free to skip this section.)

Reviewer 3: ```suggestion
(If you're not interested in the internals of the Rust compiler, feel free to skip this section.)
```

The error above came because of a pass called [`everybody_loops`]. A compiler 'pass' is a transformation on the source code, for example [finding items without documentation][missing_docs].

> The `everybody_loops` pass turns the above code into

Reviewer 3: ```suggestion
The `everybody_loops` pass turns the above code into:
```
